### PR TITLE
[Merged by Bors] - Add simple collision sound to breakout

### LINKED ISSUE #4326: Add audio to Breakout example
Just add one simple sound, for when the ball collides.

This is useful for:

1. Demonstrating that we have audio.
2. Demonstrating the very basics of using audio.
3. Testing that your audio works.

@IceSentry expressed interest in tackling this, so coordinate if you want to tackle this in the next couple of days.

### PR DESCRIPTION
**Pull request by IceSentry**
# Objective

- Add sound effect to the breakout example

## Solution

- Add a collision event and a system that listens to the event and plays a sound

I only added a single sound for all collisions for the sake of simplicity, but this could easily be extended to play a different sound depending on the type of entity hit.

The sound was generated randomly by using https://sfxr.me

https://sfxr.me/#11111GA9soYREjtsWhzjPrpMDEYSjX8Fo1E6PnKhxdw6tu869XW4EAc3nzpKVAYLMzToNcHQtQjeBqjZukqPmMDToGdYQQCWBnC3nEYfp53se5ep9btxRdLK

Closes #4326

https://user-images.githubusercontent.com/8348954/160154727-00e30743-3385-4c2f-97f0-1aaaf9a4dcc5.mp4

For some reason the video has a lot of delay in the sound, but when playing it locally there's no delay

---

## Changelog

- Added sound to breakout example
- Added bevy_audio and vorbis to the feature list ran for examples in CI

## Migration Guide

N/A

### CODE REVIEW

File: examples/game/breakout.rs [hunk 1]
```diff
@@ -167,6 +175,10 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
     commands.spawn_bundle(OrthographicCameraBundle::new_2d());
     commands.spawn_bundle(UiCameraBundle::default());
 
+    // Sound
+    let brick_collision_sound = asset_server.load("sounds/breakout_collision.ogg");
```
Comment (alice-i-cecile) [inline examples/game/breakout.rs:179]:
```suggestion
    let ball_collision_sound = asset_server.load("sounds/breakout_collision.ogg");
```

This sounds plays on collisions with the wall and paddle too.

File: examples/game/breakout.rs [hunk 2]
```diff
@@ -167,6 +175,10 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
...
+    commands.insert_resource(CollisionSound(brick_collision_sound));
```
Comment (alice-i-cecile) [inline examples/game/breakout.rs:180]:
```suggestion
    commands.insert_resource(CollisionSound(ball_collision_sound));
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Just the naming nits; otherwise I'm very happy with this.

I like that this shows off Events too: it's a really common and powerful tool that was missing from this introductory example.

Largest barrier will be to get CI to cooperate of course.

File: examples/game/breakout.rs [hunk 3]
```diff
@@ -58,12 +58,14 @@ fn main() {
...
         .add_startup_system(setup)
+        .add_event::<CollideEvent>()
         .add_system_set(
             SystemSet::new()
                 .with_run_criteria(FixedTimestep::step(TIME_STEP as f64))
                 .with_system(check_for_collisions)
                 .with_system(move_paddle.before(check_for_collisions))
-                .with_system(apply_velocity.before(check_for_collisions)),
+                .with_system(apply_velocity.before(check_for_collisions))
+                .with_system(collision_sound_player.after(check_for_collisions)),
```
Comment (alice-i-cecile) [inline examples/game/breakout.rs:68]:
```suggestion
                .with_system(play_collision_sound.after(check_for_collisions)),
```

File: examples/game/breakout.rs [hunk 4]
```diff
@@ -58,12 +58,14 @@ fn main() {
         .insert_resource(Scoreboard { score: 0 })
         .insert_resource(ClearColor(BACKGROUND_COLOR))
         .add_startup_system(setup)
+        .add_event::<CollideEvent>()
```
Comment (alice-i-cecile) [inline examples/game/breakout.rs:61]:
```suggestion
        .add_event::<CollisionEvent>()
```

File: examples/game/breakout.rs [hunk 5]
```diff
@@ -82,9 +84,15 @@ struct Velocity(Vec2);
 #[derive(Component)]
 struct Collider;
 
+#[derive(Default)]
+struct CollideEvent;
```
Comment (alice-i-cecile) [inline examples/game/breakout.rs:88]:
```suggestion
struct CollisionEvent;
```

**Review by alice-i-cecile — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@cart I'm happy with this example change from a learning perspective, but I want to get your sign-off on running this in CI.

I think it's a solid idea, because it verifies that our audio is working (we had no other functional tests for this), but it does add some complexity and a bit of time to the process.

**Comment by alice-i-cecile** [comment]
> I'm not sure it adds that much time, it seems to be around 9-10 minutes in all the recent runs, not just my PR.

Excellent, that's very good to hear.

> As for complexity, it's just a few more features in the feature list, it doesn't seem really complex to me. Are you talking about something else?

My concern is mostly that it's another set of dependencies that we can fail to fetch, or otherwise go out of date. Upon reflection though, all of the failures I've seen there just seem to be the server that hosts the packages being unresponsive. And if  something genuinely breaks externally it's better to know early 🤔 

Okay, you've thoroughly convinced me. I'm still going to wait on Cart here, but I'm on board.

File: examples/game/breakout.rs [hunk 6]
```diff
@@ -394,3 +410,14 @@ fn check_for_collisions(
         }
     }
 }
+
+fn play_collision_sound(
+    mut collide_events: EventReader<CollisionEvent>,
+    audio: Res<Audio>,
+    sound: Res<CollisionSound>,
+) {
+    // Plays a sound for each collision events
```
Comment (rparrett) [inline examples/game/breakout.rs:419]:
```suggestion
    // Plays a sound for each collision event
```

**Comment by rparrett** [comment]
The only thing I noticed while testing is that the ball can hit two adjacent bricks in the same frame and queue up two sounds to be played.

Since this is a sound that happens every time the ball bounces (rather than every time it breaks a brick), should we play the sound only once per frame if a collision occurs?

Like

```rust
// Play a sound once per frame if a collision occurred. `count` consumes the
// events, preventing them from triggering a sound on the next frame.
if collide_events.iter().count() > 0 {
    info!("playing sound!");

    audio.play(sound.0.clone());
}
```

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -274,13 +274,13 @@ jobs:
...
         run: |
-          cargo build --no-default-features --features "bevy_dynamic_plugin,bevy_gilrs,bevy_gltf,bevy_winit,render,png,hdr,x11,bevy_ci_testing,trace,trace_chrome"
+          cargo build --no-default-features --features "bevy_dynamic_plugin,bevy_gilrs,bevy_gltf,bevy_winit,render,png,hdr,x11,bevy_ci_testing,trace,trace_chrome,bevy_audio,vorbis"
       - name: Run examples
         run: |
           for example in .github/example-run/*.ron; do
             example_name=`basename $example .ron`
             echo "running $example_name - "`date`
-            time CI_TESTING_CONFIG=$example xvfb-run cargo run --example $example_name --no-default-features --features "bevy_dynamic_plugin,bevy_gilrs,bevy_gltf,bevy_winit,render,png,hdr,x11,bevy_ci_testing,trace,trace_chrome"
+            time CI_TESTING_CONFIG=$example xvfb-run cargo run --example $example_name --no-default-features --features "bevy_dynamic_plugin,bevy_gilrs,bevy_gltf,bevy_winit,render,png,hdr,x11,bevy_ci_testing,trace,trace_chrome,bevy_audio,vorbis"
```
Comment (DJMcNab) [inline .github/workflows/ci.yml:283]: What does the website do when running examples? Do we need a companion PR to there?

File: examples/game/breakout.rs [hunk 8]
```diff
@@ -82,9 +84,15 @@ struct Velocity(Vec2);
 #[derive(Component)]
 struct Collider;
 
+#[derive(Default)]
+struct CollisionEvent;
+
 #[derive(Component)]
 struct Brick;
 
+#[derive(Component)]
```
Comment (DJMcNab) [inline examples/game/breakout.rs:93]: Why is this `Component`?

**Review by DJMcNab — APPROVED** [review]
Code seems good to me, just a few points of clarification.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (rparrett) [reply .github/workflows/ci.yml:283]: I believe the website builds with default features https://github.com/bevyengine/bevy-website/blob/master/generate-wasm-examples/generate_wasm_examples.sh#L32

File: examples/game/breakout.rs (hunk 8 above)
Comment (IceSentry) [reply examples/game/breakout.rs:93]: I think it was just a copy paste mistake. It's not needed, I'll remove it as soon as I have access to a computer.

**Review by rparrett — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
bors retry

**Pull request closed without merging**